File: NODE_JS/blockchain.js
const crypto = require('crypto');

class Block {
  constructor(index, timestamp, data, previousHash = '') {
    this.index = index;
    this.timestamp = timestamp;
    this.data = data;
    this.previousHash = previousHash;
    this.hash = this.calculateHash();
  }

  calculateHash() {
    return crypto.createHash('sha256').update(
      this.index + this.previousHash + this.timestamp + JSON.stringify(this.data)
    ).digest('hex');
  }
}

class Blockchain {
  constructor() {
    this.chain = [this.createGenesisBlock()];
  }

  createGenesisBlock() {
    return new Block(0, Date.now(), "Genesis Block", "0");
  }

  getLatestBlock() {
    return this.chain[this.chain.length - 1];
  }

  addBlock(newBlock) {
    newBlock.previousHash = this.getLatestBlock().hash;
    newBlock.hash = newBlock.calculateHash();
    this.chain.push(newBlock);

    console.log("New block added to the blockchain:");
    console.log(`Index: ${newBlock.index}`);
    console.log(`Timestamp: ${newBlock.timestamp}`);
    console.log(`Data: ${JSON.stringify(newBlock.data)}`);
    console.log(`Previous Hash: ${newBlock.previousHash}`);
    console.log(`Hash: ${newBlock.hash}`);
  }

  isChainValid() {
    for (let i = 1; i < this.chain.length; i++) {
      const currentBlock = this.chain[i];
      const previousBlock = this.chain[i - 1];

      if (currentBlock.hash !== currentBlock.calculateHash()) {
        return false;
      }

      if (currentBlock.previousHash !== previousBlock.hash) {
        return false;
      }
    }
    return true;
  }

  printBlockchain() {
    console.log("Blockchain:");
    this.chain.forEach((block) => {
      console.log(`Index: ${block.index}`);
      console.log(`Timestamp: ${block.timestamp}`);
      console.log(`Data: ${JSON.stringify(block.data)}`);
      console.log(`Previous Hash: ${block.previousHash}`);
      console.log(`Hash: ${block.hash}`);
      console.log("----------------------------");
    });
  }
}

module.exports = { Blockchain, Block };

File: SPRING_BOOT/iot_security/src/main/java/com/blockchain/iot_security/blockchain/Blockchain.java
package com.blockchain.iot_security.blockchain;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class Blockchain {

    public static class Block {
        public int index;
        private final long timestamp;
        private final String data;
        private final String previousHash;
        private final String hash;

        public Block(int index, long timestamp, String data, String previousHash) {
            this.index = index;
            this.timestamp = timestamp;
            this.data = data;
            this.previousHash = previousHash;
            this.hash = calculateHash();
        }

        public String calculateHash() {
            return Integer.toHexString((index + Long.toString(timestamp) + data + previousHash).hashCode());
        }

        public String getData() {
            return data;
        }

        public String getHash() {
            return hash;
        }

        public String getPreviousHash() {
            return previousHash;
        }
    }

    private final List<Block> chain;

    public Blockchain() {
        chain = new ArrayList<>();
        chain.add(createGenesisBlock());
    }

    private Block createGenesisBlock() {
        return new Block(0, new Date().getTime(), "Genesis Block", "0");
    }

    public Block getLatestBlock() {
        return chain.getLast();
    }

    public void addBlock(String data) {
        Block latestBlock = getLatestBlock();
        Block newBlock = new Block(chain.size(), new Date().getTime(), data, latestBlock.getHash());
        chain.add(newBlock);
    }

    public List<Block> getChain() {
        return chain;
    }
}
